python/pyspark/sql/types.py

```diff
@@ -1688,7 +1688,10 @@ def _check_series_convert_date(series, data_type):
     :param series: pandas.Series
     :param data_type: a Spark data type for the series
     """
-    if type(data_type) == DateType:
+    import pyarrow
+    from distutils.version import LooseVersion
+    # As of Arrow 0.12.0, date_as_objects is True by default, see ARROW-3910
+    if LooseVersion(pyarrow.__version__) < LooseVersion("0.12.0") and type(data_type) == DateType:
```

Comment:
Yup, it will be called a lot (but not per record at least but per batch), and I also think it should be called once ideally.

I roughly guess that this has been done in this way so far because I guess we're not sure about the versions in worker side and driver side .. For instance, both versions in both codes can be different as far as I know because we don't have a check for it (correct me if I am mistaken).

Probably we should add a check like we do for Python version check between driver and worker, and have few global checks. Of course, we could do it separately I guess.

Reply:
BTW, I was thinking about targeting to upgrade minimum PyArrow version in Spark 3.0.0 since the codes are being complicated for those if-else.

Reply:
> Probably we should add a check like we do for Python version check between driver and worker, and have few global checks. Of course, we could do it separately I guess.

Yeah, we could do this but it might not really be too big of deal. I think eventually it will be sort of like Pandas versions, if they are close there will probably be no issues. But with major versions might not be completely compatible.